src-tauri/src/aux_data_structures/timeline.rs

```diff
@@ -0,0 +1,247 @@
+use crate::graph::graph_ds::Graph;
+
+// Created at the beginning of each rescue attempt.
+pub struct Timeline {
+    curr_snapshot: Option<Graph>,
+    label: i32,
+    curr_timeline_id: i32,
+    curr_snapshot_id: i32,
+    label_dir: String,
+    data_dir: String,
+}
+
+impl Timeline {
+    /// Creates a new timeline.
+    ///
+    /// # Arguments
+    ///
+    /// * `config_fields` - a HashMap containing the config fields: timeline_data_dir (location of data/timelines) and timeline_label_dir (location of data)
+    ///
+    /// # Returns
+    ///
+    /// * `Timeline` - a new timeline
+    pub fn new(config_fields: HashMap<&str, &str>) -> Timeline {
+        let mut curr_timeline_id = 0;
+
+        let timeline_data_dir = config_fields.get("timeline_data_dir").unwrap();
+        let timeline_label_dir = config_fields.get("timeline_label_dir").unwrap();
+
+        let paths = fs::read_dir(timeline_data_dir).unwrap();
+        for path in paths {
+            let path = path.unwrap().path();
+            let filename = path.file_name().unwrap().to_str().unwrap();
+            let filename = filename.split('.').collect::<Vec<&str>>()[0];
+            let filename = filename.parse::<i32>().unwrap();
+            if filename > curr_timeline_id {
+                curr_timeline_id = filename;
+            };
+        }
+
+        curr_timeline_id += 1;
+
+        Timeline {
+            curr_snapshot: None,
+            label: 1,
+            curr_timeline_id,
+            curr_snapshot_id: 0,
+            label_dir: timeline_label_dir.to_string(),
+            data_dir: timeline_data_dir.to_string(),
+        }
+    }
+
+    pub fn add_snapshot(&mut self, snapshot: Graph) {
+        match &self.curr_snapshot {
+            None => {
+                self.curr_snapshot = Some(snapshot);
+            }
+            Some(_curr_snapshot) => {
+                let is_connected = self.check_connection(&snapshot);
+                self.write_snapshot();
+                self.curr_snapshot = Some(snapshot);
+                if !is_connected {
+                    self.label = 0;
+                    self.write_timeline_label();
+
+                    self.curr_timeline_id += 1;
+                }
+            }
+        }
+        self.curr_snapshot_id += 1;
+    }
+
+    pub fn check_connection(&mut self, other_snapshot: &Graph) -> bool {
+        match &self.curr_snapshot {
+            None => true,
+            Some(curr_snapshot) => {
+                let g1_order = curr_snapshot.get_order();
+                let g2_order = other_snapshot.get_order();
+                let diff = (g1_order as i32 - g2_order as i32).abs();
+                if diff >= 1 {
+                    return false;
+                }
+                true
+            }
+        }
+    }
+
+    pub fn get_curr_snapshot(&self) -> Option<Graph> {
+        match self.curr_snapshot {
+            None => None,
+            Some(ref curr_snapshot) => Some(curr_snapshot.clone()),
```

Comment:
Is there a reason we're duplicating the current snapshot? This would mean that it wouldn't be possible to edit the snapshot being received. This cloning might be something we leave to the consumer of this function